SRUC programmers challenge 3


	One of the main challenges of using genotypes for genomic evaluations is to ensure data quality and validating that genotypes belong to the right animals. When an animal and its offspring are genotyped we can do parentage verification. When we get negative result we need to run parentage discovery.  This is very computationally demanding for large datasets. There are 108 Single Nucleotide Polymorphisms (SNPs) identified for cattle which give high accuracy for checking parent-offspring relationship. Use the list of SNPs (file: parentageSNPs.txt) and find possible sires for all animals. Simulated data can be found in file mMatrixFileSimulated_54609_1000_1.dat. All animals have been genotyped using Illumina 50k V2 chip (file: Illumina50kV2map.txt).  
	
	
Requirements:
1.	Input file names should be possible to change without altering the code.
2.	Use functions and/or procedures.
3.	Use UML naming convention.
Deliverables:


1.	Provide a flowchart of your solution. Please do not start coding until we finalise it during our next meeting. You can use http://www.gliffy.com/ (week 1)
2.	Write software which takes input files in fixed format and outputs a list of possible sires for every animal. (week 2)

